crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,471 @@
+        entity: Entity,
+        table_row: TableRow,
+    ) -> bool {
+        Self::fetch(fetch, entity, table_row)
+    }
+
+    #[inline]
+    fn update_component_access(state: &Self::State, access: &mut FilteredAccess<ComponentId>) {
+        <&Handle<A>>::update_component_access(&state.handle_id, access);
+        access.add_read(state.resource_id);
+    }
+
+    #[inline]
+    fn update_archetype_component_access(
+        state: &Self::State,
+        archetype: &Archetype,
+        access: &mut Access<ArchetypeComponentId>,
+    ) {
+        access.add_read(state.archetype_id);
+        <&Handle<A>>::update_archetype_component_access(&state.handle_id, archetype, access);
+    }
+}
+
+/// SAFETY: read-only access
+unsafe impl<A: Asset> ReadOnlyWorldQuery for AssetChanged<A> {}
+
+#[cfg(test)]
+mod tests {
+    use crate::{self as bevy_asset, AssetPlugin};
+
+    use crate::{AssetApp, Assets};
+    use bevy_app::{App, AppExit, Last, Startup, Update};
+    use bevy_core::TaskPoolPlugin;
+    use bevy_ecs::{
+        entity::Entity,
+        event::EventWriter,
+        system::{Commands, IntoSystem, Local, Query, Res, ResMut, Resource},
+    };
+    use bevy_reflect::TypePath;
+
+    use super::*;
+
+    #[derive(Asset, TypePath, Debug)]
+    struct MyAsset(usize, &'static str);
+
+    fn run_app<Marker>(system: impl IntoSystem<(), (), Marker>) {
+        let mut app = App::new();
+        app.add_plugins((TaskPoolPlugin::default(), AssetPlugin::default()))
+            .init_asset::<MyAsset>()
+            .add_systems(Update, system);
+        app.update();
+    }
+
+    #[test]
+    #[should_panic]
+    fn handle_filter_with_other_query_panics() {
+        fn compatible_filter(
+            _query1: Query<&mut Handle<MyAsset>>,
+            _query2: Query<Entity, AssetChanged<MyAsset>>,
+        ) {
+            println!(
+                "this line should not run, as &mut Handle<MyAsset> conflicts with AssetChanged<MyAsset>",
+            );
+        }
+        run_app(compatible_filter);
+    }
+
+    #[test]
+    #[should_panic]
+    fn handle_query_pos_panics() {
+        fn incompatible_query(_query: Query<(AssetChanged<MyAsset>, &mut Handle<MyAsset>)>) {
+            println!(
+                "this line should not run, as &mut Handle<MyAsset> conflicts with AssetChanged<MyAsset>",
+            );
+        }
+        run_app(incompatible_query);
+    }
+
+    #[test]
+    #[should_panic]
+    fn handle_other_query_panics() {
+        fn incompatible_params(
+            _query1: Query<&mut Handle<MyAsset>>,
+            _query2: Query<AssetChanged<MyAsset>>,
+        ) {
+            println!(
+                "this line should not run, as AssetChanged<MyAsset> conflicts with &mut Handle<MyAsset>",
+            );
+        }
+        run_app(incompatible_params);
+    }
+
+    // According to a comment in QueryState::new in bevy_ecs, components on filter
+    // position shouldn't conflict with components on query position.
```

Comment:
This is correct, but the `WorldQuery::filter_fetch` safety comment needs to be updated to mention mutual exclusion with the return value of `WorldQuery::fetch` .